Create AddLogLevelGuard Class to add guard the log statements

bhavyasripenumarthi opened the pull request:
<!--
Thank you for taking the time to contribute to OpenRewrite!
Feel free to delete any sections that don't apply to your pull request.
-->

## What's changed?
Added a new file called add log level guard to add log level guards to secure if statements
## What's your motivation?
To surround log statements by adding log level guards
## Anything in particular you'd like reviewers to focus on?
To foucs on file AddLogLevelGuard.java
## Anyone you would like to review specifically?
@timtebeek 

## Have you considered any alternatives or workarounds?
I am new to this, so I am trying it.

## Any additional context
<!-- Any thoughts you would like to share in addition to the above. -->

### Checklist
- [x] I've added the license header to any new files through `./gradlew licenseFormat`
- [x] I've used the IntelliJ auto-formatter on affected files
- [x] I've updated the documentation (if applicable)

timtebeek commented:
Thank you for opening this pull request @bhavyasripenumarthi ; that really helps!

timtebeek commented:
I've gone ahead and added some tests to you pull request, such that you can start stepping through with the debugger. That ought to help you iterate more quickly.

I suggest you have a look at the documentation we have on using the JavaTemplate, to insert an if statement where there is now a logging statement.
https://docs.openrewrite.org/authoring-recipes/modifying-methods-with-javatemplate
https://docs.openrewrite.org/concepts-explanations/javatemplate

bhavyasripenumarthi commented:
> https://docs.openrewrite.org/concepts-explanations/javatemplate

Sure, I am modifying the test now, once done will push it again.

bhavyasripenumarthi commented:
> I've gone ahead and added some tests to you pull request, such that you can start stepping through with the debugger. That ought to help you iterate more quickly.
> 
> I suggest you have a look at the documentation we have on using the JavaTemplate, to insert an if statement where there is now a logging statement. https://docs.openrewrite.org/authoring-recipes/modifying-methods-with-javatemplate https://docs.openrewrite.org/concepts-explanations/javatemplate

I modified my. code, but the tests are failing. The exception is java.lang.AssertionError: [org.example.AddLogLevelGuard display name should not end with a period.]

timtebeek commented:
You'll want to do a `git pull` as I had fixed that in https://github.com/bhavyasripenumarthi/rewrite-logging-frameworks/commit/e587dec9df83ef8a2d32ff8efb9c1324f214c3a5 pushed to your fork.

bhavyasripenumarthi commented:
> You'll want to do a `git pull` as I had fixed that in [bhavyasripenumarthi@e587dec](https://github.com/bhavyasripenumarthi/rewrite-logging-frameworks/commit/e587dec9df83ef8a2d32ff8efb9c1324f214c3a5) pushed to your fork.

Yes, I pulled that and I got the test file.

bhavyasripenumarthi commented:
Hi, i tried to run the recipe without using any build tool, by following the documentation here. 
https://docs.openrewrite.org/running-recipes/running-rewrite-without-build-tool-plugins

I am getting error at this line. If I paster the same code and made necessary imports.
List<J.CompilationUnit> cus = javaParser.parse(sourcePaths, projectDir, ctx);
Required type:
List<CompilationUnit>
Provided:
Stream
<SourceFile>

joanvr commented:
> Hi, i tried to run the recipe without using any build tool, by following the documentation here. https://docs.openrewrite.org/running-recipes/running-rewrite-without-build-tool-plugins
> 
> I am getting error at this line. If I paster the same code and made necessary imports. List<J.CompilationUnit> cus = javaParser.parse(sourcePaths, projectDir, ctx); Required type: List Provided: Stream

Yes, we need to update this documentation. The return type of `JavaParser::parser` have changed from `List<J.CompilationUnit>` to `Stream<SourceFile>`

Anyway, I'm curious on why you would like to execute the recipe without any build tool? It's not a very common use case

bhavyasripenumarthi commented:
Its a common use in my project, but somehow, I am unable to run the recipe with gradle as build tool. I am just testing it with the help of Junit only as of. when I run ./gradlew rewriteRun it was showing me Running Active recipe AddLogLevelGuard, but the change is not getting applied on my example class. So I planned to test it without any build tool.

bhavyasripenumarthi commented:
@timtebeek The recipe is ignoring the classes that were annotated with @Slf4j annotation and not making any changes in the file even if the log statements were present in that. Any idea, on how can i include them.I added the lombock package in the imports section but no luck

joanvr commented on src/main/java/org/openrewrite/java/logging/slf4j/AddLogLevelGuard.java:
```diff
@@ -0,0 +1,64 @@
+                    return mi;
+                }
+                // Prevent nested if statements
+                if (getCursor().firstEnclosing(J.If.class) != null) {
```
What if we are in another random if block? Maybe we should check the expression on the if too?

timtebeek commented:
> @timtebeek The recipe is ignoring the classes that were annotated with @slf4j annotation and not making any changes in the file even if the log statements were present in that. Any idea, on how can i include them.I added the lombock package in the imports section but no luck

We don't yet support fields generated by Lombok; what you could do is instead look to see if the class is annotated with `lombok.extern.slf4j.Slf4j` and then proceed as you would otherwise.

timtebeek commented:
Thanks for the effort so far @bhavyasripenumarthi ! Seeing how the recipe works out in practice though I have quite some concerns about being able to pull this off in practice. I've reported them on the issue in https://github.com/openrewrite/rewrite-logging-frameworks/issues/111#issuecomment-1870142462 . So while we're thankful of your efforts, sometimes it's better not to make any changes just yet, given the complexity of doing so without surprising users.
